# TriSocieties-Amplicon-Workshop
By Mara Cloutier

This tutorial and associated resources were created for a virtual workshop at the ASA-SSSA-CSA international meeting in 2020.

In this tutorial, you will learn to use DADA2 to quality filter and process amplicon sequencing reads, you will then learn how to create a phylogenetic tree in R, and perform downstream analyses such as create barplots of taxa, alpha-diverstiy, beta-diversity, and differential abundance analyses.
Here are the specific steps that you will need to perform in order to complete the tutorial (This assumes that you already have R and R Studio installed):
1. Download the .fastq.gz files, metadata file, and Amplicon-Workshop.R script provided above
2. Download the silva_nr99_v138_train_set.fa reference database file here: https://zenodo.org/record/3986799#.X24ykC2z1Bw
3. Open R Studio and open the R Script
4. Start working from the R Script
5. Play video as you go through the script: https://www.youtube.com/watch?v=MotrsdzCeVQ&feature=youtu.be

Please note that if you working with an old version of R Studio, then there may be issues with package installation. This tutorial was created using R version 4.0.2 and R Studio 1.3.1073.

Before starting the analysis on your own samples consider the questions/hypotheses you are addressing. The methods/analyses here may give you a good starting point but may not necessarily be the *best* method(s) for your data. 

email me if you are having problems or if you need clarificatin when working through the tutorial at muc345@psu.edu

Additionally, there is a powerpoint presentation here without audio. If you would like to see/listen to the one with audio you can find it here: https://www.youtube.com/channel/UC--SKXVUG-jBTLY9j517XlQ?view_as=subscriber

