# TriSocieties-Amplicon-Workshop
By Mara Cloutier

In this tutorial, you will learn to use DADA2 to quality filter and process amplicon sequencing reads, you will then learn how to create a phylogenetic tree in R, and perform downstream analyses such as create barplots of taxa, alpha-diverstiy, beta-diversity, and differential abundance analyses.
Here are the specific steps that you will need to perform in order to complete the tutorial (This assumes that you already have R and R Studio installed):
1. Download the .fastq.gz files, metadata file, and Amplicon-Workshop.R script provided above
2. Download the silva_nr99_v138_train_set.fa reference database file here: https://zenodo.org/record/3986799#.X24ykC2z1Bw
3. Create a folder to house the files for this workshop
4. Add the .fastq.gz files, meatdata file, R script, and silva database to the newly created folder
5. Open R Studio and open the R Script
6. Set the working directory to wherever the folder is that you created in Step 5
7. Start working fro the R Script

Please note that if you working with an old version of R Studio, then there may be issues with package installation.

